### PR TITLE
mxcubcore adaptation

### PR DESCRIPTION
**Pull request by marcus-oscarsson**
- Improved existing App object and renamed it to MXCUBEApplication
- Introduced MXCUBECore that replace the old blcontrol module
- Cleanup initialization and start up
- Attempt to cleanup route handling

### CODE REVIEW

File: mxcube3-server [hunk 1]
```diff
@@ -25,6 +25,6 @@ if __name__ == "__main__":
         print("No Redis server is running, exiting")
         sys.exit(1)
 
-    from mxcube3 import server, socketio
+    from mxcube3 import main
 
-    socketio.run(server, host="0.0.0.0", port=8081)
+    main()
```
Comment (marcus-oscarsson) [inline mxcube3-server:30]: Simply moved the server start to a main function cleaning up the initialization

File: mxcube3/__init__.py [hunk 2]
```diff
@@ -23,134 +23,82 @@
...
-sys.path.insert(0, fname)
-
-from mxcubecore import HardwareRepository as hwr
-
-hwr.add_hardware_objects_dirs([os.path.join(fname, "HardwareObjects")])
-
-import app as mxcube
-from config import Config
+from mxcube3.config import Config
+from mxcube3.app import MXCUBEApplication
```
Comment (marcus-oscarsson) [inline mxcube3/__init__.py:27]: Introduced a MXCUBEApplication object that encapsulates (or represents) "application" wide data and functions

File: mxcube3/__init__.py [hunk 3]
```diff
@@ -23,134 +23,82 @@
...
+from mxcube3.server import Server
```
Comment (marcus-oscarsson) [inline mxcube3/__init__.py:28]: Introduced a Server object that encapsulates flask, flask_socketio and a few other things that are server specific

File: mxcube3/routes/beamline.py [hunk 4]
```diff
@@ -1,149 +1,147 @@
...
 import logging
 
-from flask import Response, jsonify, request, make_response
+from flask import Blueprint, Response, jsonify, request, make_response
 
-from mxcube3 import server
-from mxcube3 import blcontrol
 from mxcube3.core import beamlineutils
 
+def init_routes(mxcube):
```
Comment (marcus-oscarsson) [inline mxcube3/routes/beamline.py:9]: attempt to isolate the routes and make them less dependent on a global import, please let me know what you think about this ?

File: mxcube3/app.py [hunk 5]
```diff
@@ -2,288 +2,400 @@
...
+    beamline = None
+    # XMLRPCServer
+    actions = None
+    # Plotting
+    plotting = None
 
-# Initial file list for user, initialized at login, for creating automatic
-# run numbers
-INITIAL_FILE_LIST = []
+    @staticmethod
```
Comment (meguiraun) [inline mxcube3/app.py:39]: why static method? (as opposed to adding the initialization in the __init__)

File: mxcube3/app.py [hunk 6]
```diff
@@ -2,288 +2,400 @@
...
+
+        with open("stored-mxcube-session.json", "w") as fp:
+            json.dump(data, fp)
+
+    @staticmethod
+    def load_settings():
+        """
+        Loads application wide variables from "stored-mxcube-session.json"
+        """
+        with open("stored-mxcube-session.json", "r") as f:
```
Comment (meguiraun) [inline mxcube3/app.py:381]: should not this be better in redis?

**Comment by meguiraun** [comment]
I cannot mount a sample:
```
  File "/Users/mikegu/MXCUBE/mxcubecore_upstream/mxcubecore/HardwareObjects/abstract/sample_changer/Container.py", line 221, in get_sample_address
    return str(basket_number) + ":" + "%02d" % (sample_number)
TypeError: %d format: a number is required, not str
```
casting `sample_number` to int does the job

File: mxcube3/__init__.py (hunk 2 above)
Comment (meguiraun) [reply mxcube3/__init__.py:27]: its a neat what of separating domains. I like it

File: mxcube3/routes/beamline.py (hunk 4 above)
Comment (meguiraun) [reply mxcube3/routes/beamline.py:9]: are you suggesting to do the same for all? I mean completing following method?
``` 
    def register_routes(mxcube):
        from mxcube3.routes.beamline import init_routes as init_beamline_routes
        Server.FLASK.register_blueprint(init_beamline_routes(mxcube))
```

**Comment by beteva** [comment]
> ```
>   File "/Users/mikegu/MXCUBE/mxcubecore_upstream/mxcubecore/HardwareObjects/abstract/sample_changer/Container.py", line 221, in get_sample_address
>     return str(basket_number) + ":" + "%02d" % (sample_number)
> ```
This has been the code from the very beginning, but indeed we should cast the sample_number to int.

File: mxcube3/routes/beamline.py (hunk 4 above)
Comment (marcus-oscarsson) [reply mxcube3/routes/beamline.py:9]: Yes, I just started with one route as we might need to discuss before we arrive to something final

File: mxcube3/app.py (hunk 5 above)
Comment (marcus-oscarsson) [reply mxcube3/app.py:39]: I wanted singleton like behavior for these classes (But I'm not certain we actually need a singleton). So I created a static class for this an the variables that were previously global are this class variables and all methods static (as there are no instance)

**Comment by marcus-oscarsson** [comment]
Thanks for having a look :)

The above should probably go on mxcubecore, and might also be a problem with the SampleChangerMockup because this code works with the or sample changers. We have to make sure that it works with both :)

File: mxcube3/routes/beamline.py (hunk 4 above)
Comment (meguiraun) [reply mxcube3/routes/beamline.py:9]: its a good start

**Comment by meguiraun** [comment]
thanks for this huge work!

**Pull request merged**